File: src/carve/_selection.py
"""Model selection utilities for CARVE results."""

import inspect
import warnings
from typing import Any

import numpy as np
import pandas as pd
from sklearn.base import ClusterMixin

from ._types import GridSpec, Measure, Rule

# Maps short measure aliases to canonical column names in estimator_results_
MEASURE_MAP = {
    "s": "ari_stability",
    "stab": "ari_stability",
    "stability": "ari_stability",
    "ari_stability": "ari_stability",
    "g": "ari_generalizability",
    "gen": "ari_generalizability",
    "generalizability": "ari_generalizability",
    "ari_generalizability": "ari_generalizability",
    "avg": "ari_average",
    "average": "ari_average",
    "ari_average": "ari_average",
    "pac": "consensus_pac_stability",
    "consensus_pac_stability": "consensus_pac_stability",
    "gini": "consensus_gini_stability",
    "consensus_gini_stability": "consensus_gini_stability",
    "ce": "consensus_ce_stability",
    "consensus_ce_stability": "consensus_ce_stability",
    "misclass": "misclassification_generalizability",
    "misclassification": "misclassification_generalizability",
    "misclassification_generalizability": "misclassification_generalizability",
}


def select_best_row_by_rule(
    results_df: pd.DataFrame,
    measure: Measure,
    rule: Rule,
    return_idx: bool = False,
) -> pd.Series:
    """Select the best row according to a selection rule.

    Parameters
    ----------
    results_df : pandas.DataFrame
        Results table with metric columns.
    measure : Measure
        Metric key (see ``MEASURE_MAP``).
    rule : Rule
        Selection rule. ``"max"`` picks the row with the highest score.
        ``"1se"`` picks the largest *k* within one standard error of the
        best score. ``"quantile"`` picks the largest *k* within the best
        score's quantile bounds.
    return_idx : bool, default=False
        If True, return the selected row index instead of the row itself.

    Returns
    -------
    row : pandas.Series or int
        Selected row (or row index if ``return_idx=True``).
    """
    y_col = MEASURE_MAP[measure]
    se_col = f"{y_col}_se"
    has_se = se_col in results_df.columns
    quant_col = f"{y_col}_upper"
    has_quant = quant_col in results_df.columns

    if rule == "1se" and not has_se:
        warnings.warn(
            f"Column {se_col!r} not found; falling back to 'max' rule.",
            RuntimeWarning,
            stacklevel=2,
        )
        rule = "max"
    elif rule == "quantile" and not has_quant:
        warnings.warn(
            f"Column {quant_col!r} not found; falling back to 'max' rule.",
            RuntimeWarning,
            stacklevel=2,
        )
        rule = "max"

    if rule == "max":
        return select_best_row_max(results_df, measure=measure, return_idx=return_idx)
    if rule == "1se":
        return select_best_row_1se(results_df, measure=measure, return_idx=return_idx)
    if rule == "quantile":
        return select_best_row_quantile(
            results_df, measure=measure, return_idx=return_idx
        )

    raise ValueError(f"Unknown rule: {rule}")


def select_best_estimator(
    results_df: pd.DataFrame,
    estimator_param_grids: list[GridSpec],
    measure: Measure = "stability",
    rule: Rule = "max",
    k: int | None = None,
) -> ClusterMixin:
    """Reconstruct the best estimator from a results table.

    Parameters
    ----------
    results_df : pandas.DataFrame
        Results table.
    estimator_param_grids : list of tuple
        Estimator classes and parameter grids.
    measure : Measure, default="stability"
        Metric key used for selection. Common aliases: ``"stability"`` /
        ``"s"``, ``"generalizability"`` / ``"g"``, ``"average"`` /
        ``"avg"``, ``"pac"``, ``"gini"``, ``"ce"``,
        ``"misclassification"``.
    rule : Rule, default="max"
        Selection rule.
    k : int or None, default=None
        Optional fixed number of clusters to filter by.

    Returns
    -------
    estimator : ClusterMixin
        Instantiated estimator with parameters from the selected row.
    """
    if k is not None:
        results_df = results_df[results_df["n_clusters"] == k]

    row = select_best_row_by_rule(
        results_df, measure=measure, rule=rule, return_idx=False
    )
    return build_estimator_from_row(estimator_param_grids, row)


def select_best_k(
    results_df: pd.DataFrame,
    measure: Measure = "stability",
    rule: Rule = "max",
) -> int:
    """Select the best number of clusters from results.

    Parameters
    ----------
    results_df : pandas.DataFrame
        Results table.
    measure : Measure, default="stability"
        Metric key used for selection. Common aliases: ``"stability"`` /
        ``"s"``, ``"generalizability"`` / ``"g"``, ``"average"`` /
        ``"avg"``, ``"pac"``, ``"gini"``, ``"ce"``,
        ``"misclassification"``.
    rule : Rule, default="max"
        Selection rule.

    Returns
    -------
    k : int
        Selected number of clusters.
    """
    row = select_best_row_by_rule(
        results_df, measure=measure, rule=rule, return_idx=False
    )
    return row["n_clusters"]


def select_best_row_max(
    results_df: pd.DataFrame,
    *,
    measure: Measure = "stability",
    return_idx: bool = False,
) -> pd.Series:
    """Select the row with the maximum score for a measure.

    Parameters
    ----------
    results_df : pandas.DataFrame
        Results table.
    measure : Measure, default="stability"
        Metric key.
    return_idx : bool, default=False
        If True, return the row index instead of the row.

    Returns
    -------
    row : pandas.Series or int
        Selected row (or row index if ``return_idx=True``).
    """
    measure_col = MEASURE_MAP[measure]

    if return_idx:
        return results_df[measure_col].idxmax()

    return results_df.loc[results_df[measure_col].idxmax()]


def select_best_row_1se(
    results_df: pd.DataFrame,
    *,
    measure: Measure = "stability",
    return_idx: bool = False,
) -> pd.Series:
    """Select the model with the largest k within 1 SE of the best score.

    Parameters
    ----------
    results_df : pandas.DataFrame
        Results table.
    measure : Measure, default="stability"
        Metric key.
    return_idx : bool, default=False
        If True, return the row index instead of the row.

    Returns
    -------
    row : pandas.Series or int
        Selected row (or row index if ``return_idx=True``).
    """
    if measure not in MEASURE_MAP:
        raise ValueError(f"Invalid measure {measure!r}. Options: {list(MEASURE_MAP)}")
    measure_col = MEASURE_MAP[measure]

    best_row = results_df.loc[results_df[measure_col].idxmax()]
    best_score = best_row[measure_col]
    se = best_row[f"{measure_col}_se"]
    threshold = best_score - se

    within_1se = results_df[results_df[measure_col] >= threshold]

    if return_idx:
        return within_1se["n_clusters"].idxmax()

    return within_1se.loc[within_1se["n_clusters"].idxmax()]


def select_best_row_quantile(
    results_df: pd.DataFrame,
    *,
    measure: Measure = "stability",
    return_idx: bool = False,
) -> pd.Series:
    """Select the model with the largest k within the best score's quantile bounds.

    Parameters
    ----------
    results_df : pandas.DataFrame
        Results table.
    measure : Measure, default="stability"
        Metric key.
    return_idx : bool, default=False
        If True, return the row index instead of the row.

    Returns
    -------
    row : pandas.Series or int
        Selected row (or row index if ``return_idx=True``).
    """
    if measure not in MEASURE_MAP:
        raise ValueError(f"Invalid measure {measure!r}. Options: {list(MEASURE_MAP)}")
    measure_col = MEASURE_MAP[measure]

    best_row = results_df.loc[results_df[measure_col].idxmax()]
    threshold_upper = best_row[f"{measure_col}_upper"]
    threshold_lower = best_row[f"{measure_col}_lower"]

    within_quantiles = results_df.loc[
        (results_df[measure_col] >= threshold_lower)
        & (results_df[measure_col] <= threshold_upper)
    ]

    if within_quantiles.empty:
        warnings.warn(
            "No estimators within quantile thresholds; falling back to max.",
            RuntimeWarning,
            stacklevel=2,
        )
        if return_idx:
            return results_df[measure_col].idxmax()
        return results_df.loc[results_df[measure_col].idxmax()]

    if return_idx:
        return within_quantiles["n_clusters"].idxmax()

    return within_quantiles.loc[within_quantiles["n_clusters"].idxmax()]


def build_estimator_from_row(
    estimator_param_grids: list[GridSpec],
    row: pd.Series,
) -> ClusterMixin:
    """Instantiate an estimator from a results table row.

    Parameters
    ----------
    estimator_param_grids : list of tuple
        Estimator classes and parameter grids.
    row : pandas.Series
        Selected configuration row.

    Returns
    -------
    estimator : ClusterMixin
        Instantiated estimator.
    """
    est_name = row["estimator"]

    est_class = next(
        cls for cls, _ in estimator_param_grids if cls.__name__ == est_name
    )

    valid_keys = get_estimator_param_names(est_class)
    params = row_to_estimator_params(row, valid_keys)

    return est_class(**params)


def get_estimator_param_names(est_class: type[ClusterMixin]) -> set[str]:
    """Return valid parameter names for an estimator class.

    Parameters
    ----------
    est_class : type
        Estimator class.

    Returns
    -------
    names : set of str
        Valid parameter names.
    """
    try:
        est = est_class()
        return set(est.get_params(deep=False).keys())

    except Exception:
        # Fallback: inspect the __init__ signature
        sig = inspect.signature(est_class.__init__)
        names = {
            p.name
            for p in sig.parameters.values()
            if p.name != "self"
            and p.kind
            in (inspect.Parameter.POSITIONAL_OR_KEYWORD, inspect.Parameter.KEYWORD_ONLY)
        }

        return names


def row_to_estimator_params(
    row: pd.Series,
    valid_keys: set[str],
) -> dict[str, Any]:
    """Extract estimator parameters from a results row.

    Parameters
    ----------
    row : pandas.Series
        Selected configuration row.
    valid_keys : set of str
        Parameter names allowed for the estimator.

    Returns
    -------
    params : dict
        Parameter dictionary with None/NaN filtered out.
    """
    params = {}

    for k in valid_keys:
        if k in row.index:
            v = row[k]

            if v is None:
                continue
            if isinstance(v, float) and np.isnan(v):
                continue

            params[k] = v

    return params

File: carve-r/R/selection.R
# Mirrors _selection.py. The MEASURE_MAP gives short aliases for the canonical
# metric column names found in `estimator_results_`.

MEASURE_MAP <- c(
  "s" = "ari_stability",
  "stab" = "ari_stability",
  "stability" = "ari_stability",
  "ari_stability" = "ari_stability",
  "g" = "ari_generalizability",
  "gen" = "ari_generalizability",
  "generalizability" = "ari_generalizability",
  "ari_generalizability" = "ari_generalizability",
  "avg" = "ari_average",
  "average" = "ari_average",
  "ari_average" = "ari_average",
  "pac" = "consensus_pac_stability",
  "consensus_pac_stability" = "consensus_pac_stability",
  "gini" = "consensus_gini_stability",
  "consensus_gini_stability" = "consensus_gini_stability",
  "ce" = "consensus_ce_stability",
  "consensus_ce_stability" = "consensus_ce_stability",
  "acc" = "accuracy_generalizability",
  "accuracy" = "accuracy_generalizability",
  "accuracy_generalizability" = "accuracy_generalizability"
)


.resolve_measure <- function(measure) {
  if (!measure %in% names(MEASURE_MAP)) {
    stop(sprintf(
      "Invalid measure '%s'. Options: %s",
      measure,
      paste(names(MEASURE_MAP), collapse = ", ")
    ), call. = FALSE)
  }
  unname(MEASURE_MAP[[measure]])
}


#' Select the best row of a CARVE results table by a selection rule
#'
#' Mirrors `carve._selection.select_best_row_by_rule`. Dispatches to
#' [select_best_row_max()], [select_best_row_1se()], or
#' [select_best_row_quantile()].
#'
#' @param results_df A data frame or tibble produced by CARVE's `fit()`. Must
#'   have a `n_clusters` column and the metric column corresponding to
#'   `measure` (plus `_se` / `_upper` / `_lower` columns for 1se / quantile).
#' @param measure Metric key. See [MEASURE_MAP] for accepted aliases.
#' @param rule One of `"max"`, `"1se"`, or `"quantile"`.
#' @param return_idx Logical; if `TRUE` return the selected row's integer
#'   position instead of the row itself.
#' @param not_two Logical; if `TRUE`, rows with `n_clusters == 2` are excluded
#'   from selection.
#'
#' @return A one-row data frame (or the integer position if `return_idx`).
#'
#' @export
select_best_row_by_rule <- function(results_df,
                                    measure,
                                    rule = c("max", "1se", "quantile"),
                                    return_idx = FALSE,
                                    not_two = FALSE) {
  rule <- match.arg(rule)
  if (not_two) {
    kept <- results_df$n_clusters != 2
    results_df <- results_df[kept, , drop = FALSE]
    if (nrow(results_df) == 0L) {
      stop("No configurations remain after excluding k=2.", call. = FALSE)
    }
  }

  y_col <- .resolve_measure(measure)
  se_col <- paste0(y_col, "_se")
  up_col <- paste0(y_col, "_upper")

  if (rule == "1se" && !(se_col %in% names(results_df))) {
    warning(sprintf("Column '%s' not found; falling back to 'max' rule.", se_col),
            call. = FALSE)
    rule <- "max"
  } else if (rule == "quantile" && !(up_col %in% names(results_df))) {
    warning(sprintf("Column '%s' not found; falling back to 'max' rule.", up_col),
            call. = FALSE)
    rule <- "max"
  }

  switch(
    rule,
    "max" = select_best_row_max(results_df, measure = measure, return_idx = return_idx),
    "1se" = select_best_row_1se(results_df, measure = measure, return_idx = return_idx),
    "quantile" = select_best_row_quantile(results_df, measure = measure, return_idx = return_idx)
  )
}


#' Select the best number of clusters from a CARVE results table
#'
#' Convenience wrapper that returns `n_clusters` from the row selected by
#' [select_best_row_by_rule()].
#'
#' @inheritParams select_best_row_by_rule
#'
#' @return Integer number of clusters.
#'
#' @export
select_best_k <- function(results_df,
                          measure = "stability",
                          rule = c("max", "1se", "quantile"),
                          not_two = FALSE) {
  rule <- match.arg(rule)
  row <- select_best_row_by_rule(
    results_df, measure = measure, rule = rule, return_idx = FALSE, not_two = not_two
  )
  as.integer(row$n_clusters)
}


#' Select the best estimator configuration as a list
#'
#' Returns the selected row's estimator name and parameter set. Estimator
#' instantiation (mirroring Python's `build_estimator_from_row`) is deferred
#' to the R6 `CARVE$get_estimator()` method in later phases.
#'
#' @inheritParams select_best_row_by_rule
#' @param k Optional integer. If supplied, filter `results_df` to that
#'   `n_clusters` before selection.
#'
#' @return A list with components `estimator` (character) and `params` (named
#'   list of hyperparameters).
#'
#' @export
select_best_estimator <- function(results_df,
                                  measure = "stability",
                                  rule = c("max", "1se", "quantile"),
                                  k = NULL,
                                  not_two = FALSE) {
  rule <- match.arg(rule)
  if (!is.null(k)) {
    results_df <- results_df[results_df$n_clusters == k, , drop = FALSE]
    if (nrow(results_df) == 0L) {
      stop(sprintf("No configurations with n_clusters == %d.", k), call. = FALSE)
    }
  }
  row <- select_best_row_by_rule(
    results_df, measure = measure, rule = rule, return_idx = FALSE, not_two = not_two
  )
  est_name <- as.character(row$estimator)
  # Strip metric / bookkeeping columns to recover the pure hyperparameter set.
  metric_cols <- c(
    "estimator", "n_clusters",
    outer(
      c("ari_stability", "ari_generalizability", "ari_average",
        "consensus_pac_stability", "consensus_gini_stability",
        "consensus_ce_stability", "accuracy_generalizability"),
      c("", "_se", "_upper", "_lower"),
      paste0
    )
  )
  param_cols <- setdiff(names(row), metric_cols)
  params <- as.list(row[, param_cols, drop = FALSE])
  # Drop NA entries (parameters not applicable to this estimator type).
  params <- params[!vapply(params, function(v) {
    length(v) == 1L && (is.na(v) || is.null(v))
  }, logical(1))]
  list(estimator = est_name, params = params, n_clusters = as.integer(row$n_clusters))
}


#' Select the row with the maximum score for a measure
#'
#' @inheritParams select_best_row_by_rule
#'
#' @return A one-row data frame (or integer position if `return_idx`).
#'
#' @export
select_best_row_max <- function(results_df, measure = "stability", return_idx = FALSE) {
  col <- .resolve_measure(measure)
  idx <- which.max(results_df[[col]])
  if (return_idx) return(idx)
  results_df[idx, , drop = FALSE]
}


#' Select the largest-k row within one standard error of the best score
#'
#' @inheritParams select_best_row_by_rule
#'
#' @return A one-row data frame (or integer position if `return_idx`).
#'
#' @export
select_best_row_1se <- function(results_df, measure = "stability", return_idx = FALSE) {
  col <- .resolve_measure(measure)
  best_i <- which.max(results_df[[col]])
  best_score <- results_df[[col]][best_i]
  se <- results_df[[paste0(col, "_se")]][best_i]
  threshold <- best_score - se

  keep <- which(results_df[[col]] >= threshold)
  ks <- results_df$n_clusters[keep]
  chosen <- keep[which.max(ks)]

  if (return_idx) return(chosen)
  results_df[chosen, , drop = FALSE]
}


#' Select the largest-k row within the best score's quantile bounds
#'
#' @inheritParams select_best_row_by_rule
#'
#' @return A one-row data frame (or integer position if `return_idx`).
#'
#' @export
select_best_row_quantile <- function(results_df, measure = "stability", return_idx = FALSE) {
  col <- .resolve_measure(measure)
  best_i <- which.max(results_df[[col]])
  up <- results_df[[paste0(col, "_upper")]][best_i]
  lo <- results_df[[paste0(col, "_lower")]][best_i]

  keep <- which(results_df[[col]] >= lo & results_df[[col]] <= up)
  if (length(keep) == 0L) {
    warning("No estimators within quantile thresholds; falling back to max.",
            call. = FALSE)
    if (return_idx) return(best_i)
    return(results_df[best_i, , drop = FALSE])
  }

  ks <- results_df$n_clusters[keep]
  chosen <- keep[which.max(ks)]
  if (return_idx) return(chosen)
  results_df[chosen, , drop = FALSE]
}
